### PR TITLE
TGUI for Techfabs II: The Great Recategorizing (AND ICONS) (AND MECHFABS) (AND AUTOLATHES)

### LINKED ISSUE #67758: Integrated circuit components printed in the component printer/module printer cost twice as much than from an un upgraded circuit printer
## Reproduction:
When using the component printer, all cost for components are a flat 1000 glass units compared to an unupgraded circuit printer which has 500 glass units. Along with the science protolathe being able to print circuit shells at a discount as well. Component printers are unupgradable as well meaning the circuit printer will get a further discount while the component printer will be stuck using half a sheet of glass. Module printers also use the flat 1000 glass price per component, meaning it is costly to reproduce many modules if they are sizable.

### PR DESCRIPTION
**Pull request by scriptis**
### HEY KID YOU WANT SOME ICONS? HOW ABOUT SOME CATEGORIES? IVE GOT ICONS AND NEW CATEGORIES

https://user-images.githubusercontent.com/12202230/190893973-07bcf1e6-4a58-424f-b4f9-e76e03be68ea.mp4

### I ALSO HAVE ICONS IN MECHFABS AND NEW CATEGORIES

https://user-images.githubusercontent.com/12202230/190894084-41217e85-447d-493c-a6db-03deec16acf1.mp4

### DO YOU OR A LOVED ONE SUFFER FROM AUTOLATHE CATEGORIES NOT EVEN LISTING EVERY ITEM YOU CAN PRINT IN AN AUTOLATHE? WELL DON'T WORRY I DID THAT TOO

https://user-images.githubusercontent.com/12202230/190894204-4663ee5d-3a7a-4a84-8652-4646a92e95c6.mp4

### SERIOUS TALK

- I recategorized **EVERY `/datum/design/` IN THE GAME** to be more UX friendly and **I HATE MYSELF FOR IT**
- I refactored techfab UI to **WORK ANYWHERE** for **ANY MACHINE THAT USES `/datum/design` as a SET OF MODULAR COMPONENTS**
- I moved a lot of **DESIGNS EXCLUSIVE TO THE AUTOLATHE** to also work **IN PROTOLATHES**
- I made **MATERIAL ICONS** animate between **ICON STATES** for **STACKS**
- **I PUT ICONS IN ALL OF YOUR FABRICATORS**
- I SOMEHOW DID ALL OF THIS **WITHOUT LOSING ANY PERFORMANCE**
- ALSO SUPPORTS COMPONENT PRINTERS AND MODULE DUPLICATORS

Other garbage:

- Fixed numerous spelling and consistency issues in designs
- Removed `Machine Design (<x>)` and `Computer Design (<x>)` from all relevant designs
- All designs are now in title case
- Numerous designs that were formerly autolathe exclusives can now also be printed at a protolathe (but not all); this is mostly just service equipment like drinking glasses and plates and silverware
- Circuits components can no longer be printed at a circuit imprinter (fixes #67758)
- Designs that are not sensible for a department to have are no longer accessible to that department (read: medbay printing turbine parts)

## Why It's Good For The Game

Improved UX for techfabs, but also for mechfabs and autolathes, and oh look it's pretty!

also I spent like eight hours doing nothing but categorizing `/datum/design`s and I'll cry if some version of this doesn't get merged eventually

## Changelog

:cl:
refactor: mechfabs, autolathes, component printers, and module duplicators now use techfab tgui components
refactor: every single design is now categorized and subcategorized
refactor: mechfabs and autolathes are now in typescript
qol: techfabs now have icons for what you're about to print
qol: techfab material icons are now animated
qol: techfab material icons now fade when no materials are available
qol: techfab searching no longer lags like hell
qol: techfab searching now searches all recipes instead of just the current category
qol: techfabs now have subcategorization (stock part users rejoice)
qol: techfabs now announce when new recipes are available
qol: numerous other techfab ui tweaks
balance: some designs that were formerly autolathe exclusive can now be printed at some departmental techfabs

/:cl:

### CODE REVIEW

**Comment by newfren** [comment]
what autolathe exclusive designs can now be found in techfabs, and which techfabs can they be found in?

**Comment by scriptis** [comment]
> what autolathe exclusive designs can now be found in techfabs, and which techfabs can they be found in?

- fishing rods, balloons, station bounced radios, headsets, cameras, balloons, light bulbs and tubes, all kitchen equipment, toner, pet carriers, sticky tape, plastic armblades, cap guns and ammo, chisels, spraycans, wirebrushes (service department)
- circuit tiles (science & engineering)
- flashlights (any department)
- condensers, health sensors, holodisks, ignites, proximity sensors, remote signaling devices, timers, voice analyzers (science & engineering)
- custom vendor refills, airlock, floor, and tile painters, newscaster, intercom, solar panel, status display, turret control, and camera frames (engineering)
- fluiducts (science, medbay, and service)

I also renamed _X + Y alloy here_ to actually be plasmaglass/plasteel/plastitanium/&c

**Comment by newfren** [comment]
cargo can also print rcds and could use the matter cartridges to fuel them.

**Comment by calliophis-maculiceps** [comment]
rebelnred-cum

**Comment by MrDoomBringer** [comment]
Heroic PR

**Comment by Wallemations** [comment]
Does this also have a condensed mode with no icons or just what's shown here

**Comment by scriptis** [comment]
> Does this also have a condensed mode with no icons or just what's shown here

I'll put this on my agenda, but at the moment each row is only around 50% larger than it is with the existing compact layout.

Part of the problem might be that I made some of the windows smaller than they were before :)

File: code/__DEFINES/research/research_categories.dm [hunk 1]
```diff
@@ -1,87 +1,186 @@
...
+#define RND_SUBCATEGORY_STOCK_PARTS_MISC "/Miscellaneous Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_TURBINE "/Turbine Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_TELECOMMS "/Telecommunications Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_1 "/Tier 1 Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_2 "/Tier 2 Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_3 "/Tier 3 Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_4 "/Tier 4 Parts"
+#define RND_SUBCATEGORY_STOCK_PARTS_EXCHANGERS "/Exchange Devices"
+
+#define RND_CATEGORY_SYNDICATE "/Syndicate Designs"
```
Comment (JohnFulpWillard) [inline code/__DEFINES/research/research_categories.dm:113]: I disagree with it being named Syndicate designs, they are unlocked illegally, but they're more repurposed items. I'd prefer sticking to Illegal Designs, or just keeping it to the original categories they are in (like desynchronizer in bluespace)

File: code/game/machinery/autolathe.dm [hunk 2]
```diff
@@ -59,171 +35,195 @@
...
+
+		output += list(design_data)
+
 	return output
 
-/obj/machinery/autolathe/ui_act(action, params)
+/obj/machinery/autolathe/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/sheetmaterials),
+		get_asset_datum(/datum/asset/spritesheet/research_designs)
```
Comment (JohnFulpWillard) [inline code/game/machinery/autolathe.dm:139]:
```suggestion
		get_asset_datum(/datum/asset/spritesheet/research_designs),
```

File: code/modules/mining/lavaland/tendril_loot.dm [hunk 3]
```diff
@@ -28,8 +28,10 @@
 	modkit_design = /datum/design/unique_modkit/bounty
 
 /datum/design/unique_modkit
-	category = list(RND_CATEGORY_MINING_DESIGNS, RND_CATEGORY_CYBORG_UPGRADE_MODULES) //can't be normally obtained
-	build_type = PROTOLATHE | AWAY_LATHE | MECHFAB
+	category = list(
+		RND_CATEGORY_TOOLS + RND_SUBCATEGORY_TOOLS_PKA_MODS
```
Comment (JohnFulpWillard) [inline code/modules/mining/lavaland/tendril_loot.dm:32]: no trailing commas?

File: code/modules/research/designs.dm [hunk 4]
```diff
@@ -37,8 +37,8 @@ other types of metals and chemistry for reagents).
 	var/build_path = null
 	/// List of reagents produced by this design. Currently only supported by the biogenerator.
 	var/list/make_reagents = list()
-	/// What category this design falls under. Used for sorting in production machines, mostly the mechfab.
-	var/list/category = null
+	/// What categories this design falls under. Used for sorting in production machines, mostly the mechfab.
```
Comment (JohnFulpWillard) [inline code/modules/research/designs.dm:40]:
```suggestion
	/// What categories this design falls under. Used for sorting in production machines.
```

File: code/modules/research/designs/mechfabricator_designs.dm [hunk 5]
```diff
@@ -637,254 +774,390 @@
...
+		RND_CATEGORY_MECHFAB_HONK + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT
+	)
+	departmental_flags = DEPARTMENT_BITFLAG_SCIENCE
 
 /////////////////////////////////////////
 //////////////Borg Upgrades//////////////
 /////////////////////////////////////////
 
 /datum/design/borg_upgrade_rename
-	name = "Cyborg Upgrade (Rename Board)"
```
Comment (JohnFulpWillard) [inline code/modules/research/designs/mechfabricator_designs.dm:728]: we are we removing Cyborg Upgrade from the names here?

File: code/__DEFINES/research/research_categories.dm (hunk 1 above)
Comment (scriptis) [reply code/__DEFINES/research/research_categories.dm:113]: I agree wholeheartedly :) I'll get this in

File: code/modules/research/designs/mechfabricator_designs.dm (hunk 5 above)
Comment (scriptis) [reply code/modules/research/designs/mechfabricator_designs.dm:728]: They're only visible in "cyborg modules," so it's redundant. The printed item still has the old name. I also did this for all computer and machine boards (machine design and computer design labels are _dead_)

**Comment by scriptis** [comment]
Also of note: after this, the goal is to repath/refactor autolathes, techfabs, mechfabs, biogenerators, limb growers, the ore silo, the ORM, and other machines into some subtypes. Future plans are:

- tgui for the ore silo
- departmental ore budgets
- moving ui data generation onto `/datum/design` and `/datum/material`
- actually consistent algorithm for determining machine efficiency
- rework stacking stacking machine, stacking machine console, and sheetmeister
- also while i'm at it tgui for the destructive analyzer and experimentor because i've spent three days standing next to them

**Comment by calliophis-maculiceps** [comment]
is there any way you can make their background colour go with departments, like how the R&D console has a purple background?

**Comment by scriptis** [comment]
> is there any way you can make their background colour go with departments, like how the R&D console has a purple background?

The R&D console doesn't have a purple background, and there's a _lot_ of refactoring that'd have to happen to make this work. If we had [CSS variables](https://caniuse.com/css-variables) it'd be easy, but we probably don't for like 50% of users.

**Comment by Colovorat** [comment]
holy based

**Comment by zxaber** [comment]
The exofab feels kinda, I dunno, cluttered? I think some row bars between items in a section would help.

Also, the buttons for print and add to queue could use some color to let them stand out a bit.

**Comment by Agux909** [comment]
Btw positronic brains are missing from the interface, they can't be printed from the lathe after tech's unlocked.

**Comment by scriptis** [comment]
> Btw positronic brains are missing from the interface, they can't be printed from the lathe after tech's unlocked.

I moved them to be exclusive to the mechfab

**Comment by jaydeegonz** [comment]
> I moved them to be exclusive to the mechfab

Did you also do the same for MMIs? I needed to print one from the medbay lathe and they weren't there.
It would make sense that only the mechfab would have positronics, but I would argue MMIs have their uses in medbay, too.

**Comment by scriptis** [comment]
> > I moved them to be exclusive to the mechfab
> 
> Did you also do the same for MMIs? I needed to print one from the medbay lathe and they weren't there. It would make sense that only the mechfab would have positronics, but I would argue MMIs have their uses in medbay, too.

I did, but this is a good point. I'll put MMIs back in the medbay lathe; but for positronic brains, the only reasonable place for them to be, in my opinion, is a mechfab--they only exist to be put into borgs or exosuits.

**Comment by jaydeegonz** [comment]
Yeah I agree.

File: tgui/packages/tgui/interfaces/Autolathe.tsx [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+import { Window } from '../layouts';
+import { capitalize } from 'common/string';
+import { Design, MaterialMap } from './Fabrication/Types';
+import { DesignBrowser } from './Fabrication/DesignBrowser';
+import { classes } from 'common/react';
+import { MaterialCostSequence } from './Fabrication/MaterialCostSequence';
+import { Material } from './Fabrication/Types';
+
+type AutolatheDesign = Design & {
+  buildable: boolean;
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Autolathe.tsx:12]: Byond is unable to send booleans so technically this type is wrong, you can use a `BooleanLike` alias or `number`.

File: tgui/packages/tgui/interfaces/Fabrication/MaterialAccessBar.tsx [hunk 7]
```diff
@@ -43,15 +44,15 @@ const LABEL_FORMAT = (value: number) => formatSiUnit(value, 0);
  * the ore silo. Has pop-out docks for each material type for ejecting up to
  * fifty sheets.
  */
-export const MineralAccessBar = (props: MineralAccessBarProps, context) => {
+export const MaterialAccessBar = (props: MaterialAccessBarProps, context) => {
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Fabrication/MaterialAccessBar.tsx:47]: Was the file renamed to match? Can't tell from the phone UI.

File: tgui/packages/tgui/interfaces/Fabrication/MaterialCostSequence.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ * For a given material set that can't be printed at all, the label for
+ * offending materials is red.
+ *
+ * Otherwise, the labels are white.
+ */
+export const MaterialCostSequence = (
+  props: MaterialCostSequenceProps,
+  context
+) => {
+  let { design, amount, costMap, available, align, justify } = props;
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Fabrication/MaterialCostSequence.tsx:54]: Use const here, extract costMap into its own let.

File: tgui/packages/tgui/interfaces/Fabrication/SearchBar.tsx [hunk 9]
```diff
@@ -26,6 +27,9 @@ export type SearchBarProps = {
 export const SearchBar = (props: SearchBarProps, context) => {
   const { searchText, onSearchTextChanged, hint } = props;
 
+  // Only actually update the search query every 200 milliseconds at most.
+  const db = debounce(onSearchTextChanged, 200);
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Fabrication/SearchBar.tsx:31]: Not sure if that even works, you create a new debounced version of this function on every render.

File: tgui/packages/tgui/styles/interfaces/Fabricator.scss [hunk 10]
```diff
@@ -1,17 +1,28 @@
+@use '../base.scss';
+@use '../colors.scss';
+@use '../functions.scss' as *;
+@use '../components/Button.scss' as button;
+@use '../components/Tabs.scss' as tabs;
+@use 'sass:math';
```
Comment (stylemistake) [inline tgui/packages/tgui/styles/interfaces/Fabricator.scss:6]: Put global imports at the top.

**Review by stylemistake — APPROVED** [review]
Very big PR, nitpicked a little, but looks good in general.

File: tgui/packages/tgui/interfaces/Fabrication/MaterialAccessBar.tsx (hunk 7 above)
Comment (scriptis) [reply tgui/packages/tgui/interfaces/Fabrication/MaterialAccessBar.tsx:47]: Yep!

**Pull request merged**